### PR TITLE
Clarify that "proof" is an extension point

### PR DESCRIPTION
**Pull request by msporny**
This PR is an attempt to partially address the "before-CR" concerns raised in issue #1316. It makes the following changes to the specification:

1. Separates the "Securing Mechanisms" section from the "Proofs" extension point section.
2. Clarifies that the `proof` property is an extension point that is not specific to Data Integrity. It just so happens that Data Integrity utilizes the extension point, but that doesn't preclude non-Data Integrity approaches from using the same extension point.
3. Removes duplicated normative statements between the Securing Mechanisms section and the Conformance section. There was some hand-wavy content around securing mechanisms that we had in v1.1 because securing mechanisms weren't as strongly defined as they are now in v2.0.
4. Cleans up the text from an editorial perspective from the v1.1 text (and the heavy edits that went in over the past two months around securing mechanisms and verification.

While it might seem like we're making a bunch of normative changes in this PR, we're not. We're just moving normative text around and removing duplicated text. I believe there is one new normative SHOULD (about adding securing mechanisms to the VC Specs Directory), but by and large, this PR is largely editorial (or is supposed to be).

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-model/pull/1389.html" title="Last updated on Dec 26, 2023, 3:44 PM UTC (8c5f22a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-model/1389/edcbf7e...8c5f22a.html" title="Last updated on Dec 26, 2023, 3:44 PM UTC (8c5f22a)">Diff</a>

### CODE REVIEW

File: index.html
Comment (msporny) [inline index.html:659]: The text above is removed because it came from v1.1 and we've made great strides in v2.0 to clarify the things this language was hand waving around.

File: index.html
Comment (msporny) [inline index.html:2145]: This text wasn't deleted, it was moved up into the Securing Mechanisms section.

File: index.html
Comment (msporny) [inline index.html:603]: By moving this text from the Securing Mechanisms section to the definition of a conforming document, we make it clear that a conforming document MUST be secured. Doing that also enables us to remove a bunch of hand waving we were doing around securing mechanisms.

**Review by dlongley — APPROVED** [review]

**Comment by iherman** [comment]
(As usual, I comment on the Linked Data/RDF aspect of the specification.)

***At the moment***, the `proof` property is formally defined in the [DI specification](https://www.w3.org/TR/vc-data-integrity/#proof-sets), and the official URL for the property is `https://w3id.org/security#proof`. The current version of the VCDM only "mentions", so to say, the vocabulary term, but the term itself "belongs" to the DI spec, to its vocabulary, etc.

The proposed PR text shifts the definition. It says, in [§4.10](https://pr-preview.s3.amazonaws.com/w3c/vc-data-model/pull/1389.html#proofs):

> This specification defines the proof property for the expression of [embedded proofs](https://pr-preview.s3.amazonaws.com/w3c/vc-data-model/pull/1389.html#dfn-embedded-proof).

It goes on by making the semantics statements whereby each proof is a separate named graph, and it makes it clear that the DI usage of the term is only one of its possible usages. I believe that is in contradiction with the fact that the `proof` property is formally defined in the DI specification.

To be clear, ***I have no problem with that "shift"***, but then we have to change the VCDM-DI relationships on that issue. Specifically:

- The `proof` property must be added to the VCDM vocabulary, and that vocabulary items should refer to §4.10 of the VCDM spec for its formal definition. The current `proof` property must be removed from the DI vocabulary. This also means changing the relevant `@context` files as well.
- We have to propose modifications in the DI specification, which should then "refer" to the specification of the `proof` in the VCDM. (This would probably lead to a new CR snapshot for the DI spec. But that is all right, we can do that relatively easily.) 

I realize, however, that it may create problems with deployed specifications.

***Alternatively***, i.e., if we do not want to change the current setup of the DI spec, then, in this PR, the `proof` extension point must be formulated as ***the*** extension point for the DI approach, and if other kinds of embedded proofs are defined, they would be required to define their own version of a property, which would be _something like the `proof` property_.

**Review by decentralgabe — APPROVED** [review]
thanks for this. looks clear to me.

File: index.html
Comment (selfissued) [inline index.html:1994]: What does the syntax [[[VC-JOSE-COSE]]] [[VC-JOSE-COSE]] do?  That looks odd to me and it may be a mistake.

File: index.html
Comment (selfissued) [inline index.html:1998]: Is this also a mistake?

File: index.html
Comment (selfissued) [inline index.html:2004]:
```suggestion
Securing mechanism specifications other than the ones
referred to above may also be defined as necessary. A securing mechanism
```

File: index.html
Comment (selfissued) [inline index.html:2037]:
```suggestion
           title="Choice of securing mechanism is use-case dependent">
```

File: index.html
Comment (selfissued) [inline index.html:2057]:
```suggestion
This specification defines the `proof` property extension point for the expression of
<a href="#dfn-embedded-proof">embedded proofs</a>.
It is up to each securing mechanism whether or not to use it, and if so, how to use it.
```

File: index.html
Comment (selfissued) [inline index.html:2082]:
```suggestion
<a>verifiable presentations</a> that embed a proof in the data model MAY use
the <code>proof</code> <a>property</a>. Methods of securing <a>verifiable
```

**Review by selfissued — CHANGES_REQUESTED** [review]
I will approve this PR after my suggestions are accepted.

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-12-19-vcwg-special#section2-8) on 2023-12-19

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 2.8. Clarify that "proof" is an extension point (pr vc-data-model#1389)

_See github pull request [vc-data-model#1389](https://github.com/w3c/vc-data-model/pull/1389)._




**Brent Zundel:** Clarify that `proof` is an extension point. This PR is addressing issue 1316. I'm going to let Manu talk about it.  

_See github issue [vc-data-model#1316](https://github.com/w3c/vc-data-model/issues/1316)._




**Manu Sporny:** There was a part of #1316 that I thought we could potentially address before CR. This PR attempts to clarify that `proof` is an extension point and always has been, and DI uses it in a way that's compatible, but DI isn't the only way. Other specs can use it too.  
… The only requirement is that you specify the `type` of the `proof`.  
… The concern here is that if we make `proof` only about DI then it's bound explicitly to that and we wanted to avoid that.  
… It makes some non-normative and small normative changes to clarify that. This PR also splits the securing mechanism section into two parts, one is to talk about securing mechanisms generally.  
… We talk about envelope and embedded proofs. The separate part we create the `proof` extension section explicitly.  
… It makes it much clearer that it's non-exclusive.  
… There are some concerns that Ivan has and Mike has concerns, but we shouldn't look at this as an either-or with the PRs that Mike raised and this one. I think we can merge this one and Mike can come back in an make his edits.  
… The key split here is that to talk about securing mechanisms and their specs wholly separately from the `proof` extension point.  

**Brent Zundel:** Thank you, Manu.  

**Michael Jones:** The strange thing I ran into when doing my PR was a statement that all embedded proofs must use the `proof` property. Which seems it's overly tying the hands of people writing proof mechanisms that we haven't thought of.  
… You could easily imagine some future ZKP mechanism embedding a different identifier that's specific to the mechanism. It's fine for it to be an extension point, but it shouldn't be a mandatory one for all embedded proofs. I'll review with that hat on.  

**Manu Sporny:** Totally agree, that is exactly the normative change I was hoping we could make. We're in agreement on that.  
… The previous text forced anyone doing an embedded securing mechanism to utilize proof and we should relax that and +1 to that, Mike says he agrees.  

**Michael Jones:** Ok, we're violently agreeing again, that's good news.  

**Brent Zundel:** Makes me happy.  

</details>

File: index.html
Comment (iherman) [reply index.html:1994]: It is not a mistake; it is a respec trick.

- [[[VC-JOSE-COSE]]] extracts the exact title of the publication directly into the text
- [[VC-JOSE-COSE]] creates a reference into the "references" section, and includes a formal document reference into the latter

Though useful, it is ugly... but that is beyond our control.

File: index.html
Comment (iherman) [reply index.html:1998]: see above

File: index.html
Comment (TallTed) [reply index.html:2004]:
```suggestion
Securing mechanism specifications other than those
referred to above may also be defined, as necessary. A securing mechanism
```

File: index.html
Comment (TallTed) [inline index.html:2038]:
```suggestion
There are multiple acceptable securing mechanisms, and this specification does
```

File: index.html
Comment (TallTed) [reply index.html:2057]:
```suggestion
This specification defines the `proof` property as an extension point for the
expression of <a href="#dfn-embedded-proof">embedded proofs</a>.
It is up to each securing mechanism to specify whether or not to use it, and
if so, how it is to be used.
```

File: index.html
Comment (TallTed) [inline index.html:7092]:
```suggestion
This media type can be used for credentials secured using an
```

File: index.html
Comment (TallTed) [inline index.html:7153]:
```suggestion
This media type can be used for presentations secured using an
```

**Review by TallTed — CHANGES_REQUESTED** [review]

**Comment by msporny** [comment]
@selfissued wrote:
> I will approve this PR after my suggestions are accepted.

I have applied all of your change suggestions, as well as @TallTed's change suggestions. Requesting a re-review.

**Comment by msporny** [comment]
@iherman wrote:
> it makes it clear that the DI usage of the term is only one of its possible usages. I believe that is in contradiction with the fact that the `proof` property is formally defined in the DI specification.

When it was created, many years ago, the `proof` property existed "outside" of both the "Verifiable Credentials" and "Data Integrity" specifications (because those specifications didn't exist in their current form then). The `proof` property was first used as an extension point in VCs and then DIDs (for secured DID Documents), but it was never meant to be "owned" by the VC or DID specs. The DI spec ended up formally defining /one/ way of using the `proof` extension point, because that's the only standardization path we have, but was never intended to be the ONLY securing mechanism that would use it. Ideally, `proof` would be defined in the vocabulary document, but those aren't normative, only the specs are... so, I agree that we need to align all the specifications with these realities, and not everything is going to line up cleanly. 

Things are the way they are now because these technologies ended up being standardized in the wrong order and in ways that special-purposed technologies that were always meant to be more generalized. The other thing that complicates all of this is that we do have production deployments of this stuff on a fairly significant scale and so attempting to re-align the formalities of how this stuff is defined is challenging:

* If we change the vocabulary URLs, we break many implementations out there (and there *are* many of them today). I'll assert that `https://w3id.org/security#proof` *is* the proper URL for this value (because it was always meant to be a general solution), so it shouldn't change to being a VC vocabulary-based URL.
* The formal definition of `proof` as "a general extension point for embedded proofs" shouldn't be in the VC spec, though the use of it as a general extension point should be in the VC spec.

The best specification we have for proof is still probably the Data Integrity specification, though, that does send the wrong message (it's not just for Data Integrity usage -- it's for ANY securing mechanism that would like to embed a proof without mandating that ALL securing mechanisms use the `proof` mechanism).

> * The `proof` property must be added to the VCDM vocabulary, and that vocabulary items should refer to §4.10 of the VCDM spec for its formal definition. The current `proof` property must be removed from the DI vocabulary. This also means changing the relevant `@context` files as well.

-1 for the reasons above. We have two less than ideal places to define `proof`... VC or DI. At present, DI remains the best place to put the formal definition (even though it's not ideal) because DI is a more generalized solution than VCs. We could add language to DI to say that `proof` is a generalized extension mechanism to make the point clear. VC can formalize the use of that extension mechanism for the purposes of VCs and we can have normative text in the VC spec about that (basically, this PR).

> * We have to propose modifications in the DI specification, which should then "refer" to the specification of the `proof` in the VCDM. (This would probably lead to a new CR snapshot for the DI spec. But that is all right, we can do that relatively easily.)

-1, because that makes the references backwards. The architectural layering gets all wonky. DI isn't supposed to know anything about VCs to function properly, it would be a layering violation for that to happen.

> _**Alternatively**_, i.e., if we do not want to change the current setup of the DI spec, then, in this PR, the `proof` extension point must be formulated as _**the**_ extension point for the DI approach, and if other kinds of embedded proofs are defined, they would be required to define their own version of a property, which would be _something like the `proof` property_.

Yes, we could do that. I prefer this approach. Can you please suggest some text that would adequately implement what you state above, @iherman?

**Review by decentralgabe — APPROVED** [review]

**Comment by iherman** [comment]
@msporny,

I do not want to go into the details of your comments (I do not agree with everything you wrote, but this may not be relevant at this point). I do accept (and actually anticipated) the argument that any change in the exact URL of the `proof` property may create problems with deployments, and therefore we have to adopt a pragmatic, though in my view suboptimal, approach. In other words, I agree with you on:

>> **Alternatively**, i.e., if we do not want to change the current setup of the DI spec, then, in this PR, the `proof` extension point must be formulated as the extension point for the DI approach, and if other kinds of embedded proofs are defined, they would be required to define their own version of a property, which would be _something like the `proof` property_.  
>
> Yes, we could do that. I prefer this approach. Can you please suggest some text that would adequately implement what you state above, @iherman?

Here are some thoughts that I think should be followed (the editorial details are still to be worked out, and I would trust you more than myself to find the right text). 

The core conceptual change is the following: the extension point is ***not*** the `proof` property. *The extension point is the general concept, whereby a particular embedded proof algorithm defines a property that must be used for that particular algorithm.* This specification does not specify the name or the URL of that property (it can be, e.g., `signature`, `secured`, `proof`, etc.). 

The property used by a particular embedded proof algorithm _MUST_ have the following characteristics:

- the property _MUST_ be applicable for both verifiable credentials or verifiable presentations
- the property _MUST_ relate the verifiable credential or the verifiable presentation to a separate named graph containing the details of the proof, referred to as the _proof graph_. The specification of the property must also provide, through `@context` files and corresponding vocabulary specifications, the definition of all terms that are used by that proof graph.
- the property _MUST_ secure:
  - in the verifiable credential case: the default graph of the credential
  - in the verifiable presentation case: the default graph of the presentation, all credential name graphs, as well as all proof graphs related to verifiable credentials

The [[VC-DATA-INTEGRITY]] specification, defined alongside with the VCDM specification, is a standard approach for an embedded proof, relying on the [`proof` property](https://www.w3.org/TR/vc-data-integrity/#dfn-proof); it is this algorithm (method?) that is used as an example for an embedded proof throughout this specification.

Note: it is conceivable (is it?) that, in the verifiable presentation case, a _different_ proof algorithm is used for the verifiable credentials and for the presentation itself. (If this is ***not*** a good idea, then this should be added to the bullet points above.)

We can also mention that the default context of VCDM includes the mapping of the `proof` property to make the usage of DI easier, just as it contains a number of JOSE terms to make the usage of JOSE-COSE easier.

Would that work, @msporny?

**Comment by iherman** [comment]
Just a minor layperson's question here: for enveloped proofs we do not define such a conceptual extension point; we simply refer to JOSE-COSE. I simply do not know whether there are other enveloped proof approaches out there, which may warrant a similar 'extension point' terminology. I am a little bit bothered by the asymmetry here...

**Comment by msporny** [comment]
> Would that work, @msporny?

The approach you suggest could put us back into a situation that the WG was admonished for during v1.0 and v1.1, namely, that we didn't normatively specify how to secure the payloads. The formulation you have above might create this meta-framework on what an embedded securing mechanism MUST do, but stops short in recommending any such mechanism. I'm a -1 to that potential path forward. In the v2.0 WG, we've gone to great lengths to define normative securing mechanisms.

If we were to integrate what you say, we'd go back to "not normatively specifying how to secure the payloads", which I expect would raise formal objections this time around (because we were specifically chartered to standardize securing mechanisms this time around). I expect the push-back will be: "What good is a normative verifiable credentials data model w/o at least one normative mechanism to secure the data model?" ... and that would be a fair question.

What we could do is some variation of the text that you propose (and perhaps this is what you meant by "standard approach for an embedded proof"), as further requirements for securing specifications (started in PR #1380). We could then RECOMMEND at least VC-JOSE-COSE and DI (as we do today in the spec via the normative references). That would give implementers clear normative guidance on at least two recognized ways of securing the data model without saying that other mechanisms with equivalent characteristics are not allowed.

If that works for you @iherman, I can re-formulate this PR to do what you say as long as no one else in this PR discussion objects to that direction.

> Note: it is conceivable (is it?) that, in the verifiable presentation case, a different proof algorithm is used for the verifiable credentials and for the presentation itself. (If this is not a good idea, then this should be added to the bullet points above.)

I don't think we should weigh in on whether or not that is a good idea. It's conceivable that there are presentation mechanisms that could reformulate the VCs in a way that moves the proofs to the presentation, or ones where you use selective disclosure proofs on the VCs and a non-selective disclosure proof on the presentation. There are use cases where combinations of these things are valid today and so I don't think it would be good for the group to suggest one approach or another.

@iherman wrote:
> Just a minor layperson's question here: for enveloped proofs we do not define such a conceptual extension point; we simply refer to JOSE-COSE. I simply do not know whether there are other enveloped proof approaches out there, which may warrant a similar 'extension point' terminology.

If we do the above set of changes, we could add specification requirements for "enveloped securing mechanisms" and "embedded securing mechanisms". Most of the requirements would be the same between the two (separate graphs, define all properties, etc.)

**Comment by iherman** [comment]
> > Would that work, @msporny?
> 
> The approach you suggest could put us back into a situation that the WG was admonished for during v1.0 and v1.1, namely, that we didn't normatively specify how to secure the payloads. The formulation you have above might create this meta-framework on what an embedded securing mechanism MUST do, but stops short in recommending any such mechanism. I'm a -1 to that potential path forward. In the v2.0 WG, we've gone to great lengths to define normative securing mechanisms.
> 

I am not sure that is the case. The big difference between then and now is that we (will) have a standard embedded proof mechanism defined by the WG, namely DI. Back than that was not true.

> 
> What we could do is some variation of the text that you propose (and perhaps this is what you meant by "standard approach for an embedded proof"), as further requirements for securing specifications (started in PR #1380). 

I *think* (if I understand well what you write) this is also what I meant :-)

> We could then RECOMMEND at least VC-JOSE-COSE and DI (as we do today in the spec via the normative references). That would give implementers clear normative guidance on at least two recognized ways of securing the data model without saying that other mechanisms with equivalent characteristics are not allowed.

Yes, that is what I wrote above. We define the general mechanism as an extension point, without referring to the `proof`, and then we recommend the mechanism that we standardize.

Ie, somebody _MAY_ come, at some point in the future, with a different mechanism, define a property call, say, `securedBy`, and combine it with VCDM declaring that `securedBy` "plugs" into VCDM via that extension point. And that would still be fine. But we do recommend DI (and JOSE-COSE, but that is a different matter).

> 
> If that works for you @iherman, I can re-formulate this PR to do what you say as long as no one else in this PR discussion objects to that direction.

I have the impression that we violently agree.

> 
> > Note: it is conceivable (is it?) that, in the verifiable presentation case, a different proof algorithm is used for the verifiable credentials and for the presentation itself. (If this is not a good idea, then this should be added to the bullet points above.)
> 
> I don't think we should weigh in on whether or not that is a good idea. It's conceivable that there are presentation mechanisms that could reformulate the VCs in a way that moves the proofs to the presentation, or ones where you use selective disclosure proofs on the VCs and a non-selective disclosure proof on the presentation. There are use cases where combinations of these things are valid today and so I don't think it would be good for the group to suggest one approach or another.

Ok. Forget I asked :-)

> 
> @iherman wrote:
> 
> > Just a minor layperson's question here: for enveloped proofs we do not define such a conceptual extension point; we simply refer to JOSE-COSE. I simply do not know whether there are other enveloped proof approaches out there, which may warrant a similar 'extension point' terminology.
> 
> If we do the above set of changes, we could add specification requirements for "enveloped securing mechanisms" and "embedded securing mechanisms". Most of the requirements would be the same between the two (separate graphs, define all properties, etc.)

+1

**Comment by msporny** [comment]
@iherman wrote:
> I have the impression that we violently agree.

I tried re-writing this PR and it caused a huge number of issues for the other 4 PRs that depend on this PR. So, here's what I'm going to do: I have raised issue #1400 to track your concern @iherman and we'll deal w/ it in a future PR (to be completed this week). 

We'll process the 4 other PRs (that have approvals now) and then apply your suggestions on top of all the other ones that landed.

**Comment by msporny** [comment]
Normative, multiple reviews, changes requested and made, remaining issue tracked in #1400, no objections, merging.

**Pull request merged**
